Modules/_functoolsmodule.c

```diff
@@ -461,8 +650,17 @@ partial_setstate(partialobject *pto, PyObject *state)
 {
     PyObject *fn, *fnargs, *kw, *dict;
 
-    if (!PyTuple_Check(state) ||
-        !PyArg_ParseTuple(state, "OOOO", &fn, &fnargs, &kw, &dict) ||
+    if (!PyTuple_Check(state)) {
+        PyErr_SetString(PyExc_TypeError, "invalid partial state");
+        return NULL;
+    }
+    Py_ssize_t state_len = PyTuple_GET_SIZE(state);
+    if (state_len != 4) {
+        PyErr_Format(PyExc_TypeError,
+                     "expected 4 items in state, got %zd", state_len);
+        return NULL;
+    }
```

Comment:
Are these additional checks still needed?

Reply:
This is to match error message in Python implementation.

Otherwise, can remove this and change https://github.com/python/cpython/blob/6aa35f3002dda25858d47e702e750e2871e42a7c/Lib/functools.py#L328
to `TypeError("invalid partial state")`, which is less informative.

Reply:
It is okay. This error is expected to never occur, so we should not bother to make it more specific. It is also okay if the Python implementation raises IndexError or ValueError in these cases.

This is not related to this feature, so no changes are needed.